openfl/utilities/optimizers/numpy/adam_optimizer.py

```diff
@@ -0,0 +1,125 @@
+                Initial accumulator value must be >= 0.')
+        if epsilon <= 0:
+            raise ValueError(
+                f'Invalid epsilon value: {epsilon}. Epsilon avalue must be > 0.')
+
+        self.params = params
+
+        if params is None and model_interface is not None:
+            self._set_params_from_model(model_interface)
+
+        self.learning_rate = learning_rate
+        self.beta_1, self.beta_2 = betas
+        self.initial_accumulator_value = initial_accumulator_value
+        self.epsilon = epsilon
+        self.current_step: Dict[str, int] = {param_name: 0 for param_name in self.params}
+
+        self.grads_first_moment, self.grads_second_moment = {}, {}
+
+        for param_name in self.params:
+            self.grads_first_moment[param_name] = np.full_like(self.params[param_name],
+                                                               self.initial_accumulator_value)
+            self.grads_second_moment[param_name] = np.full_like(self.params[param_name],
+                                                                self.initial_accumulator_value)
+
+    def _update_first_moment(self, grad_name: str, grad: np.ndarray) -> None:
+        """Update gradients first moment."""
+        self.grads_first_moment[grad_name] = (self.beta_1
+                                              * self.grads_first_moment[grad_name]
+                                              + ((1.0 - self.beta_1) * grad))
+
+    def _update_second_moment(self, grad_name: str, grad: np.ndarray) -> None:
+        """Update gradients second moment."""
+        self.grads_second_moment[grad_name] = (self.beta_2
+                                               * self.grads_second_moment[grad_name]
+                                               + ((1.0 - self.beta_2) * grad**2))
+
+    def step(self, gradients: Dict[str, np.ndarray]) -> None:
+        """
+        Perform a single step for parameter update.
+
+        Implement Adam optimizer weights update rule.
+
+        Args:
+            gradients: Partial derivatives with respect to optimized parameters.
+        """
+        for grad_name in gradients:
+            if grad_name not in self.grads_first_moment:
+                raise KeyError(f"Key {grad_name} doesn't exist in optimized parameters")
+
+            grad = gradients[grad_name]
+
+            self._update_first_moment(grad_name, grad)
+            self._update_second_moment(grad_name, grad)
+
+            grads_first_moment_normalized = (self.grads_first_moment[grad_name]
+                                             / (1. - self.beta_1
+                                             ** (self.current_step[grad_name] + 1)))
+            grads_second_moment_normalized = (self.grads_second_moment[grad_name]
+                                              / (1. - self.beta_2
+                                              ** (self.current_step[grad_name] + 1)))
```

Comment:
Maybe make sense create normalization function with short variables names because it looks unreadable.